keps/sig-api-machinery/20190603-immutable-fields.md

```diff
@@ -0,0 +1,313 @@
+https://github.com/kubernetes/kubernetes/issues/65973. Today this is only possible
+with the unreasonable development overhead of a webhook.
+
+### Goals
+
+- extend the CRD API to be able to specify immutability for fields.
+- publish the immutability field of CRDs via OpenAPI as vendor extension.
+- verify immutability on CR update and patch requests.
+- propose a source code marker to be consumed by kubebuilder and openapi-gen.
+- the semantics of immutability must be driven by:
+  - we do not break/change old CRD persistence semantics.
+  - the user-observed equality used for immutability checks must match the equality on
+    persisted objects. I.e. if `StorageRoundtrip(object)` is the object returned by a
+    create or update call, then we want that `StorageRoundtrip(a) == StorageRoundtrip(b)`
+    is the equality used for comparing `a` and `b`. If that check fails, a request
+    is rejected because of immutability conflict.
+- the mechanism must extend to
+  - the addition of protobuf or other encodings which unify values like empty, null and undefined.
+  - the use for existing native types in order to replace complex validation code with a simple declarative marker on the types.
+  - the restriction of the equality to only map keys, but not their values.
+  - the allowance of addition and/or deletion of map key/values.
+  - the allowance of addition and/or deletion of slice items with a unique identifier field (e.g. `.name`).  
+
+### Non-Goals
+
+- The mechanism must be extensible to native types, but its implementation is optional.
+- The mechanism must be extensible to future normalization behaviours which will be
+  required to support protobuf for CRs. But this KEP does not aim at defining these
+  and hence defining a custom equality which is compatible with normalization.
+- The mechanism is not supposed to allow different orders in lists to be considered equal.
+
+## Proposal
+
+We propose
+
+1. adding a vendor extensions to CRD OpenAPI schemas named `x-kubernetes-immutable` with boolean
+   value, but only `true` is a valid value.
+2. do **strict deep-equal** comparison of those fields listed in this extension during
+   update validation in the request pipeline. If that comparison fails, the request is 
+   rejected as 400 "Bad Request".
+   
+We create another KEP to define custom normalization steps for CRs done during 
+deserialization from etcd and when receiving a request (just after pruning and defaulting).
+
+### OpenAPI extension
+
+The `x-kubernetes-immutable` vendor extension is set in the spec of fields,
+arrays and objects. It recursively applies to the whole subtree:
+
+```yaml
+properties:
+  foo:
+    type: array
+    items:
+      x-kubernetes-immutable: true
+      type: string
+```
+
+means that the `foo` array itself is mutable, but the items are not. This means that
+items can be deleted and added, but at each existing index the string values are
+immutable.
```

Comment:
Might be worth mentioning that since maps are OpenAPI arrays (with `x-kubernetes-list-type`: `map`) that this also covers the "the restriction of the equality to only map keys, but not their values" goal listed above.

Reply:
There are also real maps, e.g. annotations and labels. We avoid them most of the times.

It is probably worth to add a section about `x-kubernetes-list-type` and `x-kubernetes-list-map-keys` and how these interact with immutability.